lib/ash/actions/update/bulk.ex

```diff
@@ -3453,4 +3516,84 @@ defmodule Ash.Actions.Update.Bulk do
         Ash.Resource.Info.action(resource, action)
     end
   end
+
+  defp find_changeset_by_ref(result, changesets_by_ref, ref_metadata_key) do
+    ref_key = result.__metadata__[ref_metadata_key]
+
+    if ref_key do
+      changesets_by_ref[ref_key]
+    else
+      find_changeset_by_id_fallback(result, changesets_by_ref)
+    end
+  end
+
+  defp find_changeset_by_id_fallback(result, changesets_by_ref) do
```

Comment:
Let's remove this. Your usage of refs should authoritatively solve this. If no changeset can be found by ref we should just raise an error. And not all resources have an `id`.